### PR TITLE
feat(flex/web): add component

### PR DESCRIPTION
**Pull request by krishnaacc123**

### CODE REVIEW

File: src/atoms/Flex/Flex.stories.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { storiesOf } from '@storybook/react';
+import Flex from './Flex';
+import Text from '../Text';
+import View from '../View';
+
+storiesOf('Flex', module)
+  .addParameters({
+    component: Flex,
+  })
+  .add('default', () => (
```
Comment (dhruvdutt) [inline src/atoms/Flex/Flex.stories.js]: Add more stories using Flex props.

File: src/atoms/Flex/Flex.web.js [hunk 2]
```diff
@@ -1,3 +1,76 @@
...
+    'space-around',
+    'space-between',
+    'space-evenly',
+  ]),
+  children: PropTypes.element.isRequired,
+};
+
+Flex.defaultProps = {
+  display: 'flex',
+  flexDirection: 'row',
```
Comment (dhruvdutt) [inline src/atoms/Flex/Flex.web.js]:
```suggestion
  flexDirection: 'column',
```

Comment (dhruvdutt) [reply src/atoms/Flex/Flex.web.js]: Defaults for web and native should be same.
Web browser default is `row` but default in react-native default is `column`.

File: src/atoms/Flex/Flex.web.js [hunk 3]
```diff
@@ -1,3 +1,76 @@
...
+    justifyContent,
+    order,
+    children,
+    style,
+    ...props
+  }) =>
+    React.cloneElement(children, {
+      ...props,
+      style: { ...style, ...children.props.style },
+    }),
```
Comment (kamleshchandnani) [inline src/atoms/Flex/Flex.web.js]: Once you do the above change you can probably just do this
```suggestion
    React.cloneElement(children, props),
```

File: src/atoms/Flex/Flex.web.js [hunk 4]
```diff
@@ -1,3 +1,76 @@
...
+    flex,
+    flexBasis,
+    flexDirection,
+    flexGrow,
+    flexShrink,
+    flexWrap,
+    justifyContent,
+    order,
+    children,
+    style,
```
Comment (kamleshchandnani) [inline src/atoms/Flex/Flex.web.js]: remove this `style` from here. don't explicitly extract it. Native needs it but web doesn't

File: src/atoms/Flex/Flex.web.js [hunk 5]
```diff
@@ -1,3 +1,76 @@
...
+    ...props
+  }) =>
```
Comment (kamleshchandnani) [inline src/atoms/Flex/Flex.web.js]: Condition to have exact 1 child is missing here

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+
+import View from '../../View';
+import Flex from '../index';
+
+describe('Renders <Flex /> correctly', () => {
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: please write tests based on the convention it was discussed the other day

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  it('renders child of <Flex />(no styles) correctly', () => {
+    const { container } = render(
+      <Flex>
+        <View />
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: Tests are meant to cover all the cases and not just one

File: src/atoms/Flex/Flex.stories.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    <Flex>
+      <View>
+        <Text>Hello1</Text>
```
Comment (kamleshchandnani) [inline src/atoms/Flex/Flex.stories.js]: write some meaningful readable text.

File: src/atoms/Flex/Flex.stories.js (hunk 1 above)
Comment (kamleshchandnani) [reply src/atoms/Flex/Flex.stories.js]: Yes and probably create knobs for them

File: src/atoms/Flex/Flex.stories.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import Flex from './Flex';
+import Text from '../Text';
+import View from '../View';
+
+storiesOf('Flex', module)
+  .addParameters({
+    component: Flex,
+  })
+  .add('default', () => (
+    <Flex>
```
Comment (deorukhkarchaitanya) [inline src/atoms/Flex/Flex.stories.js]: Flex Component doesn't do anything without props other than cloning the children. You can probably do

```suggestion
    <Flex flex={1} justifyContent="space-evenly">
```

File: src/atoms/Flex/Flex.web.js [hunk 10]
```diff
@@ -1,3 +1,76 @@
...
+    React.cloneElement(children, {
+      ...props,
+      style: { ...style, ...children.props.style },
+    }),
+)`
+  &&& {
```
Comment (deorukhkarchaitanya) [inline src/atoms/Flex/Flex.web.js]: Why do you need ```&&& { .... }```

Can't we just write it the way it's on ```Flex.native.js```?

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
```
Comment (deorukhkarchaitanya) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: More tests please

File: src/atoms/Flex/Flex.web.js (hunk 5 above)
Comment (dhruvdutt) [reply src/atoms/Flex/Flex.web.js]: I think it's missing in native as well. @shubhnik Can you fix this?

File: src/atoms/Flex/Flex.web.js (hunk 10 above)
Comment (kamleshchandnani) [reply src/atoms/Flex/Flex.web.js]: Excellent Question! We can't because Css specificity issues my boi. This basically increases the specificity so no other styles can override this unless explicitly mentioned using `!important`. [Read Here](https://razorpay.slack.com/archives/CTKFJ1QVC/p1585041853007200)

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 7 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: Added more tests

File: src/atoms/Flex/Flex.stories.js (hunk 8 above)
Comment (krishnaacc123) [reply src/atoms/Flex/Flex.stories.js]: Added knob

File: src/atoms/Flex/Flex.stories.js (hunk 9 above)
Comment (krishnaacc123) [reply src/atoms/Flex/Flex.stories.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 11 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added more tests

**Comment by krishnaacc123** [comment]
I'm little confused about the test case description for nested components. Don't know if this is correct.

File: src/atoms/Flex/Flex.web.js (hunk 3 above)
Comment (krishnaacc123) [reply src/atoms/Flex/Flex.web.js]: done

File: src/atoms/Flex/Flex.web.js (hunk 4 above)
Comment (krishnaacc123) [reply src/atoms/Flex/Flex.web.js]: removed

File: src/atoms/Flex/Flex.web.js (hunk 5 above)
Comment (krishnaacc123) [reply src/atoms/Flex/Flex.web.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 6 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/Flex.stories.js (hunk 1 above)
Comment (krishnaacc123) [reply src/atoms/Flex/Flex.stories.js]: created knobs

File: src/atoms/Flex/Flex.stories.js (hunk 9 above)
Comment (kamleshchandnani) [reply src/atoms/Flex/Flex.stories.js]: @deorukhkarchaitanya it'll render with default props which is render children as row. So it should be fine

File: src/atoms/Flex/Flex.web.js [hunk 12]
```diff
@@ -1,3 +1,76 @@
...
+    'space-around',
+    'space-between',
+    'space-evenly',
+  ]),
+  children: PropTypes.element.isRequired,
+};
+
+Flex.defaultProps = {
+  display: 'flex',
+  flexDirection: 'column',
```
Comment (kamleshchandnani) [inline src/atoms/Flex/Flex.web.js]: make it row. for web it should be row because that's the default behavior of the browser

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+import { render } from '@testing-library/react';
+
+import View from '../../View';
+import Space from '../../Space';
+import Flex from '../index';
+
+describe('<Flex />', () => {
+  describe('flexDirection', () => {
+    it('renders a child with flexDirection attribute and flex properties correctly', () => {
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]:
```suggestion
    it('renders a child with flexDirection row', () => {
```

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+      const { container } = render(
+        <Flex flexDirection="row">
+          <View />
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: Make it contentful. Alone View doesn't makes sense at all

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        </Flex>,
+      );
+      expect(container).toMatchSnapshot();
+    });
+  });
+  describe('flex', () => {
+    it('renders an inline styled child with flex attribute and flex properties correctly', () => {
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]:
```suggestion
    it('renders an inline styled child with flex property', () => {
```

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+      const viewStyles = { backgroundColor: 'red', height: 100 };
+      const { container } = render(
+        <Flex flex={1}>
+          <View style={viewStyles} />
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: make it contentful

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    });
+  });
+  describe('<Sapce /> <Flex /> together plays properly', () => {
+    it('renders child of <Space />(no styles) and <Flex />(no styles) correctly', () => {
+      const { container } = render(
+        <Space>
+          <Flex>
+            <View>
+              <View as="span">Flex Item #1</View>
+              <View as="span">Flex Item #2</View>
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: do like this for even above tests

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+      const viewStyles = { backgroundColor: 'red', height: 100 };
+      const { container } = render(
+        <Flex flex={1}>
+          <View style={viewStyles} />
+        </Flex>,
+      );
+      expect(container).toMatchSnapshot();
+    });
+  });
+  describe('<Sapce /> <Flex /> together plays properly', () => {
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: this doesn't follows the convention

Comment (kamleshchandnani) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: also this shouldn't be a new test suite i.e `describe` is not required here

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    it('renders child of <Space />(no styles) and <Flex />(no styles) correctly', () => {
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: `renders child by applying flex and space to it`

File: src/atoms/Flex/Flex.web.js (hunk 10 above)
Comment (deorukhkarchaitanya) [reply src/atoms/Flex/Flex.web.js]: Nice! :O

But do we need it here? Won't it cause confusion while debugging as to why some styles might be getting overwritten? The consumer will then have to dive into the code of our DS and look into this specific place to understand the higher specificity?

Doing this on a project makes sense where it is _our_ code but on a DS it seems like a heavy enforcement, no?

Comment (kamleshchandnani) [reply src/atoms/Flex/Flex.web.js]: No it’s basically to ensure that our enhancers space, flex etc work as expected it’s not breaking anything in the consumer codebase it’s just applying the class with higher specificity for this particular style not everything

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 19 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 18 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 17 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 16 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 15 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 14 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 13 above)
Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/Flex.web.js (hunk 12 above)
Comment (krishnaacc123) [reply src/atoms/Flex/Flex.web.js]: done

File: src/atoms/Flex/Flex.stories.js [hunk 20]
```diff
@@ -0,0 +1,88 @@
+import React from 'react';
+import { storiesOf } from '@storybook/react';
+import { select, text } from '@storybook/addon-knobs';
+
```
Comment (dhruvdutt) [inline src/atoms/Flex/Flex.stories.js]: No need to add newline here.

Comment (dhruvdutt) [inline src/atoms/Flex/Flex.stories.js]:
```suggestion

```

Comment (dhruvdutt) [reply src/atoms/Flex/Flex.stories.js]: Removed this and commited.

File: src/atoms/Flex/__tests__/Flex.web.test.js [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+import { render } from '@testing-library/react';
+
+import { renderWithTheme } from '../../../_helpers/testing';
+import View from '../../View';
+import Text from '../../Text';
+import Space from '../../Space';
+import Flex from '../index';
+
+describe('<Flex />', () => {
+  it('renders child by applying flex and space to it', () => {
```
Comment (kamleshchandnani) [inline src/atoms/Flex/__tests__/Flex.web.test.js]: this just slipped off my mind sorry.
But the stories are meant to be nested when you have multiple props. Look at this for [reference](https://github.com/razorpay/blade/pull/70/files#diff-2bc383cd6dec7f8a022ce6813e94d55d)

Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: Should I nest this inside `it('as')...`

Comment (kamleshchandnani) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: add `describe('components composition',() => { it('renders child by applying flex and space properties'`

Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: Since all other test cases are having multiple components so should I move all to this `describe` ?

Comment (kamleshchandnani) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: no. just for this. because you're testing this explicitly. Rest are focussing on flex properties so that's fine

Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: oh, thanks

Comment (krishnaacc123) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: added

File: src/atoms/Flex/__tests__/Flex.web.test.js (hunk 11 above)
Comment (shubhnik) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: @krishnaacc123 we need to test if inline styles are being applied because that was one edge case. You can take reference from native tests for Flex.

Comment (shubhnik) [reply src/atoms/Flex/__tests__/Flex.web.test.js]: @krishnaacc123  we need to test if inline styles are being applied because that was one edge case. You can take reference from native tests for Flex.
